src/integrator_interface.jl

```diff
@@ -48,77 +73,49 @@ end
-    perform_step!(integrator, integrator.cache)
+        # u(tprev) is not modified hence we do not have to copy it
+        integrator.integrator.uprev = integrator.sol.u[integrator.prev2_idx]
 
-    if typeof(integrator.prob) <: ConstantLagDDEProblem
-        #warn("ConstantLagDDEProblem is deprecated. Use DDEProblem instead.")
-        neutral = false
-        constant_lags = integrator.prob.lags
-    else
-        neutral = integrator.prob.neutral
-        constant_lags = integrator.prob.constant_lags
+        # do not have to reset interpolation data in initial time step since always a
+        # constant extrapolation is used (and interpolation data of solution at initial
+        # time point is not complete!)
+        if length(integrator.sol.t) > 1
+            recursivecopy!(integrator.integrator.k, integrator.sol.k[end])
+        end
     end
 
-    # if dt is greater than the minimal lag, then use a fixed-point iteration
-    if isempty(constant_lags) ||
-       (integrator.dt > minimum(constant_lags) && isfinite(integrator.EEst))
+    # reset boolean which indicates whether history function was evaluated at a time point
+    # past the final point of the current solution
+    integrator.integrator.isout = false
 
-        # update cached error estimate of integrator
-        integrator.integrator.EEst = integrator.EEst
+    # perform always at least one calculation
+    perform_step!(integrator, integrator.cache)
 
-        # save dt, u(tprev) and interpolation data of interval [tprev, t] of ODE
-        # integrator since they are overwritten by fixed-point iteration
-        if typeof(integrator.uprev_cache) <: AbstractArray
-            recursivecopy!(integrator.uprev_cache, integrator.integrator.uprev)
-        else
-            integrator.uprev_cache = integrator.integrator.uprev
-        end
-        recursivecopy!(integrator.k_integrator_cache,
-                       view(integrator.integrator.k, 1:integrator.kshortsize))
-        integrator.integrator.dtcache = integrator.integrator.dt
-
-        # move ODE integrator to interval [t, t+dt] to use interpolation of ODE integrator
-        # in the next iterations when evaluating the history function
-        integrator.integrator.t = integrator.t + integrator.dt
-        integrator.integrator.tprev = integrator.t
-        integrator.integrator.dt = integrator.dt
-        if typeof(integrator.integrator.uprev) <: AbstractArray
-            recursivecopy!(integrator.integrator.uprev, integrator.uprev)
-        else
-            integrator.integrator.uprev = integrator.uprev
-        end
+    # if the history function was evaluated at time points past the final time point of the
+    # solution, i.e. returned extrapolated values, continue with a fixed-point iteration
+    if integrator.integrator.isout
+        # update ODE integrator to next time interval together with correct interpolation
+        advance_ode_integrator!(integrator)
```

Comment:
Don't know what this is yet.

Reply:
When the next step is calculated (at this time only once, it will continue with fixed-point iterations) we should add that information to the history function. Therefore the ODE integrator has already to move to the time interval `[t, t+dt]`, and this is done in `advance_ode_integrator!`. If the ODE integrator is not updated to the next time interval any call `h(t1)` for some `t1` in `[t, t+dt]` will still use the initial extrapolation of the ODE integrator from the last time step in `[tprev, t]`, so e.g. even in the next fixed-point iterations we would still use our initial guesses when evaluating the history function.

Reply:
I see because it doesn't auto-advance anymore. Got it.